Preprogrammed ranges info in widgets (#162)

juzen2003 opened the pull request:
- Initial implementation for #162 
- Local branches: useful_ranges_162
- Currently available in the following widgets:
  - ring radius
  - wavelength
- Some notes for behaviors of the dropdown list:
  - "Preprogrammed Ranges" button is placed below input fields. I was trying to put it in the same line with input fields, but default font sizes are different in different browsers. Without making the button smaller and widen the widget, the button will not stay in the same line with input fields in Safari and FF, and it will shrink to the next line once the widget width is getting narrower, so I make it a link button and place it below input fields (regular button in the next line looks a bit weird).
  - When dropdown list is open, and the widget column is scrolled, the dropdown list will close. This matches the behavior of string autocomplete dropdown.
  - When scrolling reaches to both ends in dropdown list and we keep scrolling (or wheeling), overscrolling will not happen (outer ps in widget column will not get scrolled in this case).
  - When dropdown list is closed, any expanded items will be collapsed. So the next time when dropdown list is open, all list items are collapsed and only categories are shown.

juzen2003 committed:
Add init steps to store ranges info in a separate file and store it in
database in param_info table.
- Create param_info_ranges.json, this is the file used to store ranges info.
- In param_info.json (line 102-105), add a new field "ranges".
- In paraminfo/modals.py (line 27), add a new column "ranges".
- In obs_ring_geometry.json (line 85), add "pi_ranges": "obs_ring_geometry"
  in "ring_radius1" field.
- In do_param_info.py (line 27-39), read the contents of param_info_ranges.json
  and convert it from a str to dict. Contents in dict will be retrieved
  and stored in database later.
- In do_param_info.py (line 71-76, 99), if "pi_ranges" exists, we will
  obtain the corresponding ranges data from param_info_ranges.json, convert
  it to str, and store in database.

juzen2003 committed:
Add a customized django template filter to convert json key string to a
prettier string for display. Add init html structure (bootstrap dropdown)
to display ranges info for obs_ring_geometry. Still need to add javascript
to see if the dropdown display looks nice.
- Create apps/ui/templatetags/prettify_str.py to store the customized
  django template filter. This is to replace underscore with a space and
  capitalize every word in a string.
- In paraminfo/models.py (line 120-131), add get_ranges_info function to
  obtain a python dict that only contains ranges info from database. Units
  and qtype are ignored.
- In ui/views.py (line 296, 309), update api_get_widget to pass in ranges
  info to widget.html.
- In ui/widget.html (line 47-67), use bootstrap dropdown to display ranges
  info. Still need to work on javascript to see if it looks good (init
  draft).

juzen2003 committed:
Add Saturn Rings info to param_info_ranges.json.
- In param_info_ranges.json (line 25-103), add Satrun Rings info about
  ring radius.

juzen2003 committed:
Add init draft of UI to display a dropdown list for useful ranges. The
dropdown list is built using bootstrap form control class, and can be
easily switched to normal select tag by removing the boostrap class. This
is a commit before switch branches to test out UI using bootstrap dropdown.
- In ui/widget.html (line 49-66), add select tag to display useful ranges
  in a dropdown list.
- In widgets.js (line 412-417), update to properly display popover for
  qtype if there is a useful ranges for the widget. This is to make sure
  useful ranges dropdown list is always at the most right hand side.

juzen2003 committed:
Add initial draft of useful ranges dropdown list (Use bootstrap dropdown
and collapse). Still need to fine tune the dropdown input location and
build up columns for dropdown list contents (Use grid system).
- In widget.html (line 50-97), add initial structure of useful ranges
  dropdown using bootstrap dropdown and collapse.
- In opus.css (line 1942-1954), add styling for useful ranges dropdown.
- In widgets.js (line 80-91), add event handler to avoid closing dropdown
  menu when clicking any dropdown item and expand/collapse info when
  clicking a dropdown submenu.

juzen2003 committed:
Make useful ranges info align in the expandable area. Use bootstrap row
and column to achieve this.
- In widget.html (line 59-78), add bootstrap row and column to display
  useful ranges info. This will help to align the info.
- In opus.css (line 1953), add styling to make sure default x scrollbar
  will not show up in the dropdown menu.

juzen2003 committed:
Update the data structure of param_info_ranges.json. Need to modify the
way we access data stored in ranges field of param_info table.
- In param_info_ranges.json (whole file), modify the data structure.
- In opus.css (line 1955-1960), widget.html (line 59-83), these modifications
  are for UI of ranges info dropdown list. Need to modify these later after
  finishing up modifying the method we access data from database.

juzen2003 committed:
Update param_info_ranges.json, remove unnecessary "," so that import can
run through.
- In param_info_ranges.json (line 189), remove the undesired "," so that
  import can run through.

juzen2003 committed:
Update the method of obtaining ranges data from database, and update the
html structure of dropdown list based on new return data structure. And
also remove unnecessary customized django filter.
- Remove unnecessary customized django filter: prettify_str.py.
- In paraminfo/models.py (line 125, 128), update get_ranges_info to convert
  string to json object (dict) so that it can be accessed in html later.
- In ui/widget.html (line 55-86), update the html structure based on new
  data structure of ranges info json.
- In opus.css (line 1962-1965), add styling to make sure each dropdown list
  div (each line) will take up the whole space of the dropdown menu.
- In param_info_ranges.json (line 100), correct the typo.

juzen2003 committed:
- Update the layout of dropdown expandable list for ranges info.
    - When button is at the right edge of the widget, dropdown expandable
      list will expand to the left to display all info.
    - When button is at the left edge of the widget (when browser is shrinking
      and the button is squeezed to the next line), dropdown expandable
      list will expand to the right to display all info.
- In widget.html (line 139-183), move dropdown expandable list out of the
  input ul and put it into a separated ul.
- In widgets.js (line 426-431), update getWidget to properly append the
  helpIcon next to qtype input.
- In opus.css (line 1954-1966, 1975-1990), update the styling to achieve
  the expected behaviors:
  (1) When button is at the right edge of the widget, dropdown expandable
    list will expand to the left to display all info.
  (2) When button is at the left edge of the widget (when browser is shrinking
    and the button is squeezed to the next line), dropdown expandable
    list will expand to the right to display all info.

juzen2003 committed:
Improve the responsiveness of ranges info dropdown expandable list. Fix
overscrolling issue. Now when scrolling reaches to both ends in dropdown
list, it will not overscroll the ps in widget container if we keep scrolling.
- In widget.html (line 142), remove bootstrap d-inline-flex, we will need
  to change this value when browser is narrower, so we put the styling
  in opus.css, that way it's easier to manage the styling changes.
- In opus.css (line 1982-1985, 1911-1995, 2002-2004), these are for better
  responsiveness of ranges info dropdown list when browser is narrower.
- In widgets.js (line line 577-581), close the dropdown when ps in widget
  container is scrolling.
- In widgets.js (line 584-588), prevent overscrolling when when scrolling
  reaches to both ends in dropdown list. Now it will not scroll the ps in
  widget container.

juzen2003 committed:
Collapse expanded contents when ranges dropdown list is closed.
- In widgets.js (line 92-96), add the event handler to collapse expanded
  contents when ranges dropdown list is closed.

juzen2003 committed:
Add arrow indicator for ranges info dropdown button and list items.
- In opus.css (line 1992-2011), add the stylings for arrow rotation when
  user clicks ranges info button or dropdown list items.
- In widget.html (line line 142-143, 149), add collapse class for dropdown
  list items. This class will be used in css styling to determine if the
  collapse item is expanded.

juzen2003 committed:
Update ranges dropdown list stylings to make sure the ranges button stay
in one line with input fields for all browsers: Chrome, FF, and Safari if
the browser width is large enough.
- In opus.css (line 1496), increase the max width of widget column, this
  will make sure widget has enough width to fit the ranges button in the
  same line with input fields. (Different browsers have different font
  sizes, by increasing the max width, we can make sure the button can stay
  at the same line with inputs when the widget is first open.)
- In opus.css (line 1985, 1989-1990), change display to inline-block.
  Previously we use inline-flex, but in FF, the button will not stay at
  the same line as input fields even if widget column width is increased,
  so we use inline-block to achieve this.

juzen2003 committed:
Add the ability to perform search when user selects one of ranges info
items from the dropdown.
- In widgets.js (line 97-132), add event handler to perform search when
  user selects one of ranges info items.
- In widget.html (line 162), store slug value in data-widget of
  .op-useful-ranges-data-item.
- In param_info_ranges.json (line 32), fix the typo value.

juzen2003 committed:
Update the layout of ranges info dropdown list, add data from csv files
for ring radius.
- In param_info_ranges.json, add data from csv files for ring radius.
- In widget.html (line 54), add bootstrap w-100 to .scrollable-menu
  so that dropdown list will shrink with the widget.
- In widget.html (line 51), use bootstrap .btn-link as the ranges info
  button. This looks better compared to regular button.
- In opus.css (line 1955), make the border of dropdown list a bit thicker
  so it looks more clear when overlapping with widgets.
- In opus.css (line 1964), remove underscore of .btn-link.

juzen2003 committed:
Add preprogrammed wavelength ranges info.
- In param_info_ranges.json (line 570-655), add preprogrammed ranges info
  for wavelength.
- In obs_wavelength.json (line 82), add "pi_ranges": "obs_wavelength" so
  that preprogrammed ranges info for wavelength will be stored in param_info
  table for obs_wavelength.

juzen2003 committed:
Add ranges info data for Jupiter inner satellites, Saturn regular satellites,
Uranus inner satellites, and Neptune inner satellites.
- In param_info_ranges.json, add ranges info for Jupiter inner satellites,
  Saturn regular satellites, Uranus inner satellites, and Neptune inner
  satellites. Data from:
  https://pds-rings.seti.org/jupiter/jupiter_tables.html
  https://pds-rings.seti.org/saturn/saturn_tables.html
  https://pds-rings.seti.org/uranus/uranus_tables.html
  https://pds-rings.seti.org/neptune/neptune_tables.html

rfrenchseti commented on opus/import/table_schemas/param_info.json:
```diff
@@ -98,6 +98,11 @@
         "field_key": "primary",
         "field_notnull": true
     },
+    {
```
This should go above "id" so "id" and "timestamp" are always the last two fields in every table.

rfrenchseti commented on opus/application/apps/ui/templates/ui/widget.html:
```diff
@@ -45,6 +46,49 @@ <h6 class="card-title mt-0">
+                                                </div>
+                                            </div>
+                                            {% for range_info in range_data_entry.ranges %}
+                                                        <div class="op-useful-ranges-data-item row" data-min="{{range_info.field1|intcomma}}" data-max="{{range_info.field2|intcomma}}" data-widget="widget__{{slug}}">
```
Fix indent amounts

rfrenchseti commented on opus/application/apps/ui/templates/ui/widget.html:
```diff
@@ -45,6 +46,49 @@ <h6 class="card-title mt-0">
                         {{ form|safe }}
                 </ul>
+                {% if ranges %}
+                        <div class="op-useful-ranges dropdown">
```
Fix indent amounts

rfrenchseti commented on opus/application/apps/ui/templates/ui/widget.html:
```diff
@@ -4,6 +4,7 @@
     #op-qtype-tooltip div is appended in base.html, and popover is initialized
     in getWidget function in widgets.js
 {% endcomment %}
+{% load humanize %}
```
Didn't we decide to get rid of this?

juzen2003 replied:
This is not the customized one, this is for filter "intcomma" to add "," in numbers.

rfrenchseti commented on opus/application/static_media/css/opus.css:
```diff
@@ -1939,3 +1940,51 @@ ul.ui-autocomplete a.ui-state-active {
 }
+
+/* Styling for ranges info dropdown */
+.op-useful-ranges .dropdown-item:active {
```
Should call the class op-preprogrammed-ranges for consistency

rfrenchseti commented on opus/application/static_media/js/widgets.js:
```diff
@@ -409,7 +462,13 @@ var o_widgets = {
                                     <i class="fas fa-info-circle"></i></a></li>';
-                    $(`#widget__${slug} .widget-main ul`).append(helpIcon);
+
+                    // if ($(`#widget__${slug} .widget-main ul .op-useful-ranges`).length !== 0) {
```
Remove comments

rfrenchseti commented on opus/application/apps/ui/templates/ui/widget.html:
```diff
@@ -45,6 +46,49 @@ <h6 class="card-title mt-0">
                 </ul>
+                {% if ranges %}
+                        <div class="op-useful-ranges dropdown">
+                            <button type="button" id="dropdownMenu" class="btn btn-link dropdown-toggle" data-toggle="dropdown" aria-haspopup="true" aria-expanded="false">
```
Rename dropdownMenu to op-ranges-dropdown-menu

rfrenchseti commented on opus/import/do_param_info.py:
```diff
@@ -54,6 +68,13 @@ def create_import_param_info_table():
+            # from dict and convert it to str before storing to database
+            ranges = column.get('pi_ranges', None)
+            if ranges:
+                ranges = ranges_json.get(ranges)
```
Use ranges_json[ranges] because we don't want to have a default.
Check to see if key is in ranges first and display an error if not.

rfrenchseti commented on opus/import/do_param_info.py:
```diff
@@ -54,6 +68,13 @@ def create_import_param_info_table():
                            f'"{category_name}/{field_name}" has RANGE type '
                            +'without numerical format')
 
+            # if pi_ranges exist in .json, get the corresponding ranges info
```
exists

rfrenchseti commented on opus/import/table_schemas/param_info_ranges.json:
```diff
@@ -0,0 +1,1091 @@
+{
+    "obs_ring_geometry": [
+        {
+            "cat_name": "jupiter rings",
```
All of these names (category and fields) should be capitalized the way we want them to display. We shouldn't re-capitalize them later, taking control away from the person who wrote this JSON file.

rfrenchseti commented on opus/import/table_schemas/param_info_ranges.json:
```diff
@@ -0,0 +1,1091 @@
+    "obs_ring_geometry": [
+        {
+            "cat_name": "jupiter rings",
+            "heading1": "min",
```
Change to "Min (km)" and "Max (km)"

rfrenchseti commented on opus/import/table_schemas/param_info_ranges.json:
```diff
@@ -0,0 +1,1091 @@
+    "obs_wavelength": [
+        {
+            "cat_name": "wavelengths",
+            "heading1": "min",
```
Change to "Min (microns)" and "Max (microns)"

rfrenchseti commented on opus/application/apps/ui/templates/ui/widget.html:
```diff
@@ -45,6 +46,49 @@ <h6 class="card-title mt-0">
+                            <ul class="dropdown-menu scrollable-menu pl-0 w-100" aria-labelledby="dropdownMenu" role="menu">
+                                {% for range_data_entry in ranges %}
+                                    <li class="mr-0 d-block">
+                                        <a class="dropdown-item bg-light dropdown-toggle collapsed" data-toggle="collapse" href="#{{range_data_entry.cat_name|slugify}}" role="button" aria-expanded="false" aria-controls="{{range_data_entry.cat_name|slugify}}">{{range_data_entry.cat_name|title}}</a>
```
Don't title case here - do it in the JSON file. See below.

rfrenchseti commented on opus/application/apps/ui/templates/ui/widget.html:
```diff
@@ -45,6 +46,49 @@ <h6 class="card-title mt-0">
+                                            {% for range_info in range_data_entry.ranges %}
+                                                        <div class="op-useful-ranges-data-item row" data-min="{{range_info.field1|intcomma}}" data-max="{{range_info.field2|intcomma}}" data-widget="widget__{{slug}}">
+                                                            <div class="col-4">
+                                                                {{range_info.name|title}}
```
Don't title case here

rfrenchseti reviewed (changes requested):
- Make the "Preprogrammed ranges" the same color and weight as the main categories on the left of the search page. Don't change color on hover.
- If there is only one category, open it automatically.
- Put the units after the min and max table headers like "Min (km)".
- Figure out how to right-justify the numbers and the table header so that we can align things by decimal point. You will need some amount of margin on the right to make it look good.
- Change the width of the whole drop-down box to be narrower. It doesn't need to be the width of the whole widget.
- For wavelength, remove Gamma Ray, X-Ray, and Submillimeter and everything later
- For wavelength, fix range of ultraviolet
- For wavelength, add other colors found at https://www.energy.gov/sites/prod/files/styles/borealis_photo_gallery_large_respondlarge/public/2018/04/f50/figure%201.jpg?itok=K1QNG90T
Use the data from column 1 (CRC)
- For ring radius, figure out what's wrong with the Dawes Ringlet and the Anthe, Methone, and Pallene rings. Also I think those last three should be called "Ring Arcs".

juzen2003 committed:
Clean up code styles in widget.html, widgets.js and rename some variables.
- In widget.html (line 50-90), fix the indentation.
- In widgets.js (line 465), remove unwanted comments.
- In opus.css (line 1956-1974), add test areas, these will be removed later.
- In opus.css, widget.html, widgets.js, rename "op-useful-ranges" to
  "op-preprogrammed-ranges", and "dropdownMenu" to "op-ranges-dropdown-menu".
- In do_param_info.py (line 71-80), modify the logic to store ranges info
  in database if the pi_ranges key exists in param_info_ranges.json, otherwise
  display an error in import process.
- In param_info_ranges.json (line 1028), fix the ultraviolet info.
- In param_info.json (line 95-99), move "ranges" field ahead of "id" and
  "timestamp".

juzen2003 committed:
Remove django title filter in widget.html and modify data in param_info_ranges
to have capitalized first letter.
- In widget.html (line 57,63,66,72), remove title django template filter.
- In param_info_ranges.json, capitalize first letter for each category and
  field name.

juzen2003 committed:
Add new wavelength data from:
https://www.energy.gov/sites/prod/files/styles/borealis_photo_gallery_large_respondlarge/public/2018/04/f50/figure%201.jpg?itok=K1QNG90T
- In param_info_ranges.json (line 1045-1087), add new wavelength data.

juzen2003 committed:
Add pi_field_hints1 and pi_field_hints2 in param_info table, these will
be used to store hints text in the input. Also capitalize "Min" and "Max"
in range inputs and make them black to only have values displayed in green.
- In param_info.json (line 99-108), add "pi_field_hints1" and "pi_field_hints2"
  in param_info table fields. These will be used to store the grey hints
  text for inputs with preprogrammed ranges.
- In param_info_ranges.json (line 1087), removed unwanted "," to avoid
  import error.
- In obs_ring_geometry.json (line 86-87) and obs_wavelength.json (line 83-84),
  add pi_field_hints1 and pi_field_hints2 values to have grey hints text
  displayed in these two widgets.
- In import/do_param_info.py (line 104-105), store hints text in param_info
  table if they exist in .json file like ones in obs_ring_geometry.json
  and obs_wavelength.json.
- In paraminfo/models.py (line 28-29), create field_hints1 and field_hints2
  in ParamInfo object.
- In ui/views.py (line 297-298, 312-313), return field_hints1 and field_hints2
  to html, these might not needed later. We still need to figure out how to
  move the dropdown expandable list to the input in django form.
- search/forms.py (line 119-131), send the grey hints text to the placeholder
  of inputs.
- In opus.css (line 2018-2035), style the placeholder to have "..." when
  it can't be fit into the input, and increase the input width to hold
  the whole placeholder text when browser is wide enough.
- In search.js (line 369-371, 542-544, 544-556), restructure the html so
  that it will be easier to style just the values of hints.
- In opus.css (line 711-720), make only values of hints to green, keep the
  descriptions "Min", "Max", "Nulls" black.

juzen2003 committed:
Rename "range_hints" to "op-range-hints"

juzen2003 committed:
Add behaviors to open preprogrammed ranges dropdown list when click on
ranges input.
- In form.py (line 129), store slug in data-slugname in input.min, this
  will be used to properly toggle the corresponding ranges dropdown.
- In widget.html (line 51-54), comment out button and move
  "#op-ranges-dropdown-menu" to the div container. When user click on
  input.min, the dropdown list will show up.
- In widgets.js (line 129-169), add addPreprogrammedRangesBehaviors function
  to store all the event handlers for ranges dropdown list. This function
  will be called in getWidget.

juzen2003 committed:
Re-structure the customized ranges info dropdown list so that the list will
always stay right below input.min. Still need to figure out the right way
to toggle dropdown.
- In forms.py (line 130-132), add required attributes for input.min to toggle
  the customized dropdown list.
- In widget.html (line 50), remove unwanted class for dropdown list.
- In opus.css (line 1948-1951), add styling to make sure ranges info div
  not splitting input min and max into two lines.
- In widgets.js (line 658-666), move the whole ".op-preprogrammed-ranges"
  into the input.min li and use it as the customized dropdown expandable
  list for input.min. This will also make sure dropdown list always stays
  below input.min.
- In widgets.js (line 139, 148), update addPreprogrammedRangesBehaviors
  (update the event handlers).

juzen2003 committed:
Close ranges info dropdown list when widget container ps is scrolling.
- In widgets.js (line 648-650), toggle the dropdown when widget container
  ps is scrolling.

juzen2003 committed:
Update param_info_ranges.json data to make sure no data is left empty.
- In param_info_ranges.json (line 373-389, 892-893, 934-935), update
  ranges data.

juzen2003 committed:
Add the logic to expand and collapse ranges info dropdown. Still need to
figure out the proper way to deal with race conditions of expand & collapse
animation since they are async. Refer to:
https://getbootstrap.com/docs/4.3/components/collapse/
- In forms.py (line 126), add dropdown-toggle class to input.min when
  ranges info is available.
- In widget.html (line 56, 70-71), store proper data- and class name to
  easily access the collapsing preprogrammed ranges items. Still under
  experiment.
- In opus.css (line 2027-2029), add op-hide-preprogrammed-ranges-item class
  to hide preprogrammed ranges items.
- In widgets.js (lien 647-652), update comments related to selector for
  toggling dropdown.
- In search.js (line 78-143), add the experiment logic to expand/collapse
  ranges info. Still need to
  figure out the proper way to deal with race conditions of expand & collapse
  animation since they are async.

juzen2003 committed:
Merge remote-tracking branch 'upstream/master' into useful_ranges_162

juzen2003 committed:
Fix the race conditions of expanding/collapsing animations.
- In widget.html (line 54), rename "scrollable-menu" to "op-scrollable-menu".
- In widget.html (line 58), put slug in data-input of .container, this is
  planned to be used to select .container's corresponding input field.
- In opus.css, rename "scrollable-menu" to "op-scrollable-menu".
- In widgets.js (line 139,148), rename "scrollable-menu" to "op-scrollable-menu".
- In search.js (line 75-97), add event handlers to collapse shown and hidden
  events. These are to avoid the race conditions of expanding/collapsing
  animations.
- In search.js (line 99-105), use "isTriggeredFromInput" to identify if
  the collapse/expand events are triggered by input or mouse click. If it's
  triggered by mouse click, we still want to keep the default behavior. If
  it's triggered by input, ranges info will show/hide automatically based
  on the count of matched characters.

juzen2003 committed:
Add the function to highlight letters of ranges names that match user's
input.
- In search.js, rename "itemsMatched" to "rangesNameMatchedCounter".
- In search.js (line 388-405), add hightlightMatchedRangesName to highlight
  letters of ranges names that match user's input.

juzen2003 committed:
Add the function to remove highlighted characters of ranges names. This
will avoid the previous highlighted letter being displayed when the current
input is empty.
- In search.js (line 407-414), add removeHighlightedRangesName to remove
  all highlighted characters of ranges names. This will avoid the previous
  highlighted letter being displayed when the current input is empty.
- In search.js (line 389), fix the typo, rename "hightlightMatchedRangesName"
  to "highlightMatchedRangesName".

juzen2003 committed:
Prevent collapsed empty categories (in the case when there is no matched
characters of ranges info from user's input) from being expanded when
user clicks it.
- In widget.html (line 58), update data-input="{{slug}}" to
  data-widget="widget__{{slug}}" so that we can use it to select the
  corresponding input.
- In search.js (line 99-108), add the event handler to show.bs.collapse
  to have e.preventDefault. This will prevent collapsed empty categories
  from being expanded when user clicks it

rfrenchseti commented on opus/import/do_param_info.py:
```diff
@@ -23,6 +24,19 @@ def create_import_param_info_table():
+            contents = fp.read()
+            ranges_json = json.loads(contents)
+        except json.decoder.JSONDecodeError:
+            impglobals.LOGGER.log('debug', f'Was reading ranges json file "{ranges_filename}"')
```
This can be "logger.log" instead of "impglobals.LOGGER.log"

rfrenchseti commented on opus/import/do_param_info.py:
```diff
@@ -54,6 +68,17 @@ def create_import_param_info_table():
+                    ranges = json.dumps(ranges)
+                else:
+                    logger.log('error',
+                               f'pi_ranges: "{ranges}" is not in "{ranges_filename}"')
```
Need to "return False" after this so the import is aborted. Same is true of previous error logging for units at line 55 that I didn't catch before.

rfrenchseti commented on opus/application/apps/search/forms.py:
```diff
@@ -114,11 +115,23 @@ def __init__(self, *args, **kwargs):
-                     label = label,
-                     widget = forms.TextInput(attrs={'class':label + ' RANGE'}),
+                    required=False,
+                    label = label.title(),
```
What is the "title()" for?

juzen2003 replied:
Change to capitalize(), this is used to change "min" & "max" to "Min" & "Max".

rfrenchseti replied:
OK, I see what's going on here - since the min/max is also a class name. That does seem kind of like a bad class name, though. How about prepending something like "op-range-input-" to the "min"/"max" to make a better class name?

rfrenchseti reviewed (changes requested):
A few comments on brief experimentation:

- If you pull up the dropdown, scroll it, leave focus, and then come back, the dropdown remembers where the scroll was before. I think it would be better to reset it to the top.
- If you start typing a range name, it should scroll to the first category where there are still range names showing so you know there is a positive result.
- When there are no results in a category, right now it's confusing because it looks like you can still expand the category by clicking on it because of the ">" arrow. It might work better to grey out the category name, and perhaps remove the ">" arrow. In fact I think I'm changing my mind about showing empty categories at all. With the long list of ring categories, it's just confusing. Perhaps we should hide the inactive categories entirely.
- When there is only one result and you hit enter, it doesn't select that result.
- When you use TAB to change focus into an input field, it doesn't pop up the dropdown.
- If the focus is in a numeric range and the dropdown is showing, and then you click on the input field a second time, the dropdown goes away.
- The wavelength colors should be in different ranges. The red/green/blue colors should be marked something like "CRC Wavelength Ranges", the others should be "PDS4 Wavelength Ranges" and the eventual Johnson filters should be marked "Johnson Filters".
- "Name" in "Radius of Feature Name" should be capitalized.

juzen2003 committed:
Add return False when there is an error log message in import pipeline.
That way the import will be aborted when error messages show up.
- In do_param_info.py (line 58, 82), add return False when there is an
  error log message. This will make sure import is aborted when error
  messages show up.
- In form.py (line 123), change label.title() to label.capitalize(). This
  is to make sure min and max is capitalized.
- In widgets.js (line 80-88), experiment codes, will be removed later.

juzen2003 committed:
Prevent the shown ranges dropdown from being closed when clicking on the
focused input once again.
- In widgets.js (line 153-165), call preventDefault in hide.bs.dropdown
  event of a shown dropdown. This will prevent the shown ranges dropdown
  from being closed when clicking on the focused input once again.
- In forms.py (line 120-123), update the SearchForm so that when ranges
  info is not available, the input will not have dropdown classes and its
  data-toggle will not be set to dropdown.

juzen2003 committed:
Update wavelength json data.
- In /table_schemas/param_info_ranges.json (line 1005-1125), update
  wavelength info in obs_wavelength. Now it has three categories:
  1. "PDS4 Wavelength Ranges"
  2. "CRC Wavelength Ranges"
  3. "Johnson Filters"
- In /table_schemas/obs_ring_geometry.json (line 86), capitalize
  text in pi_field_hints1.

juzen2003 committed:
Reset dropdown scrollbar to top when there dropdown menu is open and there
is no matched characters of ranges info.
- In search.js (line 116-120), reset dropdown scrollbar location when
  ranges dropdown menu is open.

juzen2003 committed:
Scroll ranges dropdown scrollbar to the first matched category if there
is match characters of ranges info.
- In search.js (line 435-459), add setRangesDropdownScrollbarPos function
  to set the ranges info dropdown scrollbar to the first matched category.

juzen2003 committed:
Hide the empty category (including the category collapse button) when
there is no match for ranges info.
- In search.js (line 162, 178-180), hide the empty category when there is
  no match for ranges info.
- Rename "op-hide-preprogrammed-ranges-item" to "op-hide-element".

juzen2003 committed:
Add proper behaviors to perform search when user types ranges names.
- In widgets.js (line 141-165), create a function "fillRangesInputs" to
  set the values in min/max inputs of the target widget.
- In widgets.js (line 90), call "fillRangesInputs" in the event handler
  when user clicks the ranges info item.
- In search.js (line 404-484), create function "compareInputWithRangesInfo"
  to compare user's input and ranges name.
  1. If input field is empty, display the whole list with all categories
     collapsed.
  2. If input matches any of list items, expand those categories. Highlight
     and display matched items, and hide all unmatched items. Collpase and
     hide the empty categories.
- In search.js (line 135-136), call "compareInputWithRangesInfo" in the
  event handler of user input event.
- In search.js (line 198-234), add logic in the change event handler of
  ranges input to perform the search if there is only one matched ranges
  info.

juzen2003 committed:
Make sure dropdown is not shown if user focus into an input with numerical
value.
- In search.js (line 124-131), add logic and e.preventDefault() to
  show.bs.dropdown event, and it will make sure dropdown is not shown if
  user focus into an input with numerical value.
- In widgets.js (line 106, 114), remove unwanted comments.

juzen2003 committed:
Organize customized behaviors of preprogrammed ranges dropdown and expandable
list when user types/focus in & out of the input into a function called
"addPreprogrammedRangesSearchBehaviors".
- In search.js (line 357-419), create addPreprogrammedRangesSearchBehaviors
  to store all customized event handlers for preprogrammed ranges dropdown
  and expandable list when user types/focus in & out of the input.
- In widgets.js (line 102-103), update comments.

juzen2003 committed:
Fix the issue that dropdown will not open properly when previous input is
an invalid value, and current one is valid. Also properly highlight (in
red) input value that has multiple matched ranges results when user focuses
out.
- In search.js (line 76-80), properly highlight (in red) the input when
  it has multiple matched ranges results after user focuses out.
- In search.js (line 439-443), return from compareInputWithRangesInfo when
  the ranges dropdown is not available. This will prevent unnecessary steps
  from being executed,
- In search.js (line 501-510), properly open the ranges dropdown. This will
  fix the issue that dropdown is not open when previous input is an invalid
  value, and current one is valid.

juzen2003 committed:
Properly open up ranges dropdown when user tabs to focus in the input.
- In search.js (line 64-73), add the logic in focus event handler to properly
  open up ranges dropdown when user tabs to focus in the input.

juzen2003 committed:
Properly toggle dropdown when user clicks/tabs to focus in the ranges input.
- In search.js (line 66-78), properly toggle the dropdown when user tabs
  in the input. Use o_widgets.isKeepingRangesDropdownOpen to properly handle
  the race condition for two dropdown open events. One is triggered from
  clicking into the input and the other one is triggered from focusing in
  the input.
- In widgets.js (line 21), rename "isTriggeredFromInput" to
  "isKeepingRangesDropdownOpen".

juzen2003 committed:
Add protection to trigger dropdown toggling or compare input value with
ranges info only when the focus-in input is min input with .op-scrollable-menu
available.
- In search.js (line 70, 459), add the protection to trigger dropdown toggling
  or compare input value with ranges info only when the focus-in input is
  min input with .op-scrollable-menu available.

juzen2003 committed:
Remove data of Methone ring, Anthe ring, Pallene ring, Unnamed ring, and
Arago ring in param_info_ranges.json

juzen2003 committed:
Align data in ranges info by decimal point. Update the responsiveness of
ranges info dropdown.
- In widgets.js (line 704-730), add a function "alignRangesDataByDecimalPoint"
  to align ranges data by decimal point.
- In opus.css (line 1985-2027), add the styling for dropdown responsiveness
  and ranges data alignment.
- In widget.html (line 71, 76), add .op-preprogrammed-ranges-min-data and
  .op-preprogrammed-ranges-max-data. These will be used to update the html
  text with span tags & text which will be used align ranges data.
- In param_info_ranges.json, update data in wavelength.

rfrenchseti commented on opus/application/apps/search/forms.py:
```diff
@@ -114,11 +115,27 @@ def __init__(self, *args, **kwargs):
 
                 pi = get_param_info_by_slug(slug, 'search')
+                # placeholder for input hints (only apply to Min input for now)
+                hints = pi.field_hints1 if pi.field_hints1 else ''
```
Shouldn't we go ahead and support the hints2 field even though we don't have any cases that use it yet?

rfrenchseti commented on opus/application/apps/search/forms.py:
```diff
@@ -114,11 +115,27 @@ def __init__(self, *args, **kwargs):
-                     required=False,
-                     label = label,
-                     widget = forms.TextInput(attrs={'class':label + ' RANGE'}),
+                    required=False,
```
Add space around =

rfrenchseti commented on opus/application/apps/ui/templates/ui/widget.html:
```diff
@@ -45,6 +46,46 @@ <h6 class="card-title mt-0">
+                                            <div class="col-4">
+                                            </div>
+                                            <div class="col-4 font-weight-bold text-right">
+                                                {{range_data_entry.heading1}}
```
Elsewhere we've been putting a space after {{ and before }}

rfrenchseti commented on opus/application/apps/ui/templates/ui/widget.html:
```diff
@@ -45,6 +46,46 @@ <h6 class="card-title mt-0">
+                                                </div>
+                                                <div class="col-4 op-preprogrammed-ranges-min-data">
+                                                    {% if range_info.field1 %}
+                                                        {{range_info.field1|intcomma}}
```
What happens with "intcomma" when we have non-int/float field values, like date/time strings?

rfrenchseti commented on opus/application/static_media/js/search.js:
```diff
@@ -41,41 +41,76 @@ var o_search = {
+            let slugName = $(this).data("slugname");
+            let inputToTriggerDropdown = $(`#widget__${slugName} input.min`);
+            let preprogrammedRangesDropdown = $(`#widget__${slugName} .op-scrollable-menu`);
+            if (preprogrammedRangesDropdown.length === 0 || !$(e.target).hasClass("min")) {
```
Instead of returning here, wouldn't it be better to combine the if directly below and change the condition here to length !==0 && $(e.target)....  That way it's easy to add things later to this function without worrying about it prematurely returning here.

rfrenchseti commented on opus/application/static_media/js/search.js:
```diff
@@ -41,41 +41,76 @@ var o_search = {
     slugEndpointsReqno: {},
     slugRangeInputValidValueFromLastSearch: {},
 
+    rangesNameMatchedCounter: {},
```
It seems odd to have a variable named "Counter" be a dictionary. What is this indexed by? Can you put that in the name somehow? Like rangesNameMatchedCounterBySlug or something.

rfrenchseti commented on opus/application/static_media/js/search.js:
```diff
@@ -129,10 +167,47 @@ var o_search = {
+                    }
+                }
+                let allItemsInMatchedCat = $(`${matchedCatId} .op-preprogrammed-ranges-data-item`);
+                for (const singelRangeData of allItemsInMatchedCat) {
```
single not singel

rfrenchseti commented on opus/application/static_media/js/search.js:
```diff
@@ -301,6 +376,199 @@ var o_search = {
+         */
+
+        // Make sure ranges info shows up automatically when the count of matched characters
+        // is not 0. We do this in the event handler when the collpasing is done. This is to avoid
```
collapsing

rfrenchseti commented on opus/application/static_media/js/search.js:
```diff
@@ -301,6 +376,199 @@ var o_search = {
+            $(".op-scrollable-menu").scrollTop(0);
+        });
+
+        // Make sure dropdown is not shown if user focus into an input with numerical value.
```
Is this just for numerical value or also for any string value that doesn't match?

rfrenchseti commented on opus/application/static_media/js/search.js:
```diff
@@ -301,6 +376,199 @@ var o_search = {
+            }
+            o_search.setRangesDropdownScrollbarPos(preprogrammedRangesDropdown);
+            // Note: the selector to toggle collapse should be the one with "collapse" class.
+            // $(`#${containerId}`).collapse("show");
```
What is this for?

rfrenchseti commented on opus/application/static_media/js/search.js:
```diff
@@ -41,41 +41,76 @@ var o_search = {
+    // Use to determine if we should automatically expand/collapse ranges info. If it's set
+    // to true, we will automatically expand/collapse ranges info depending on the matched letters.
+    isTriggeredFromInput: false,
+    performInputValidation: true,
```
This variable could use a comment too

rfrenchseti commented on opus/application/static_media/js/widgets.js:
```diff
@@ -18,6 +18,7 @@ var o_widgets = {
      **/
 
     lastStringSearchRequestNo: 0,
+    isKeepingRangesDropdownOpen: false,
```
I think this variable could use a comment. The name isn't clear to me.

rfrenchseti commented on opus/application/static_media/js/widgets.js:
```diff
@@ -583,12 +692,40 @@ var o_widgets = {
+        let preprogrammedRangesInfo = $(`#${widget} .op-scrollable-menu li`);
+        for (const category of preprogrammedRangesInfo) {
+            let collapsibleContainerId = $(category).data("category");
+            let rangesInfoInEachCategory = $(`#${collapsibleContainerId} .op-preprogrammed-ranges-data-item`);
```
Probably should be called rangesInfoInOneCategory

rfrenchseti reviewed (changes requested):
- Highlight colors are inconsistent. When you hover over a string completion hint you see purple, but when you hover over a range completion hint you see dark grey. I also note that the string completion hint moves a little side to side when you hover over it. I think grey is probably the right answer in both cases.
- I also noticed that if you have the string choice list open and resize the browser, the dropdown becomes detached from the input field.

juzen2003 committed:
Update the layout and responsiveness of ranges info.
- In opus.css (line 1986-2043), update the css styling for ranges info.
  Now the dropdown width will be the max-content, ranges min and max data
  will never wrap. When browser is wide, ranges names will not wrap. When
  browser is less than 950px, ranges names will start to wrap by word
  (break point 950px). When browser is less than 650px, ranges names will
  wrap by letter.
- In widgets.js (line 715-721), for ranges info with long name, we make
  it wrap by letter at the beginning. This will prevent the long name from
  overlapping with data.

juzen2003 committed:
Update the layout of ranges dropdown to have max-content as the width. And
change the unit micron to symbol.
- In opus.css (line 2026-2044), update the styling the give more spaces
  between ranges name and the min values so that the long ranges names will
  not overlap with min values before being wrapped into 2nd line.
- In param_info_ranges.json, change (microns) to (&mu;m), this will change
  microns to the mu symbol.
- In widget.html (line 60, 63), use safe filter to display mu symbols for
  wavelength ranges dropdown.

juzen2003 committed:
Update to have only one "field_hints" in param_info table. And put
"pi_field_hints": "xxx (some hints)" in obs_xxx.json for one "field_name"
if we want to have hints for that specific field.
- Remove "pi_field_hints1" and "pi_field_hints2", now we use "pi_field_hints"
  for any field that wants hints.
- In forms.py, put spaces around "=" for variable assignment.

juzen2003 committed:
Fix code styles and comments.
- In widget.html, add spaces after {{ and before }}.
- In search.js (line 189-193, 393), fix typos, change "singelRangeData"
  to "singleRangeData" and "collpasing" to "collapsing".
- In search.js (line 438-439), update the comments for the event handler.
- In search.js (line 44-50, 54-55), widgets.js (line 21), add comments for
  variables: "rangesNameMatchedCounterByCategory", "rangesNameTotalMatchedCounter",
  "performInputValidation", and "isKeepingRangesDropdownOpen".
- In search.js (line 473-476), widgets.js (line 712-714), change
  "rangesInfoInEachCategory" to "rangesInfoInOneCategory".
- In search.js (line 77-82), update the logic in range input "input" event
  handler.

juzen2003 committed:
Reorder the data in CRC Wavelength Ranges in increasing numerical order.
Rename ranges input "min" and "max" class to "op-range-input-min" and
"op-range-input-max".

juzen2003 committed:
Customize jquery autocomplete hover background color to match the
customized ranges dropdown hover background color.
- In opus.css (line 2105-2108), customize jquery autocomplete hover background
  color to grey.

juzen2003 committed:
Make sure string choice dropdown is attached right below the corresponding
input when browser is resized.
- In widgets.js (line 745-758), create attchedStringDropdownToInput to
  detect string input position and set it to jquery ui autocomplete. This
  will prevent autocomplete dropdown from detaching when browser is resized.
- In opus.js (line 635), call attchedStringDropdownToInput in browser resize
  event handler.

juzen2003 committed:
Fix the issue when focus out from an empty string input, the search will
become invalid and have "?" displayed in total result count.
- In search.js (line 610), move opus.allInputsValid = false; inside the if
  block, this will prevent empty string input from getting invalid search.

juzen2003 committed:
Remove unwanted console.log in widgets.js

juzen2003 committed:
Add an ajax callback for failed dataimages api call to remove spinner.
Reset rangesNameTotalMatchedCounter when normalize input is about to be called.
- In browse.js (line 2129-2131), remove the spinner when dataimages api
  call has error. This will be browse page blank without spinner spinning.
- In search.js (line 240), reset rangesNameTotalMatchedCounter to 0 before
  calling normalize input api.

juzen2003 committed:
Fix typo, change "attchedStringDropDownToInput" to "attachStringDropdownToInput".

juzen2003 committed:
Properly attach jquery ui autocomplete dropdown to corresponding input
and make it the same width as the input.
- In widgets.js (line 745-759), update "attachStringDropdownToInput".
  Update the calculation of autocomplete dropdown width and position.

juzen2003 committed:
Fix the width and update the layout of ranges dropdown. Make ranges data
properly align with headers.
- In widgets.js (line 769-789), create getPaddingValFromDigitsInFraction
  to get the padding-right value for ranges data based on its number of
  digits. The value will be used to properly align the ranges data and
  header.
- In widgets.js (line 715-767), update alignRangesDataByDecimalPoint to
  add padding-right values for each data item. This will properly align
  each data with headers.
- In opus.css (line 1987), set width of ranges dropdown to 500px when browser
  window is wide. This will make sure the width of ranges dropdown stays
  the same when data in any category is expanded/collapsed.

juzen2003 committed:
Remove the wiggling effect when user hovers over to menu items in jquery
autocomplete.
- In opus.css (line 2095-2100), set margin to 0 to prevent wiggling when
  user mouses over to menu items in autocomplete. (default is -1).

rfrenchseti commented on opus/application/static_media/js/search.js:
```diff
@@ -41,41 +41,82 @@ var o_search = {
+            let slugName = $(this).data("slugname");
+            let inputToTriggerDropdown = $(`#widget__${slugName} input.op-range-input-min`);
+            let preprogrammedRangesDropdown = $(`#widget__${slugName} .op-scrollable-menu`);
+            if (preprogrammedRangesDropdown.length !== 0 && $(e.target).hasClass("op-range-input-min")) {
```
Shouldn't this all just be a single if with clauses connected by &&?

rfrenchseti reviewed (changes requested):
Here are the currently remaining issues:

- Wrapping of range names happens too soon - they shouldn't wrap if there is still a lot of whitespace left on the line.
- search.js line 77 if combine into a single if
- This still happens: "If you type in something that could be a partial match (like "Thebe") and then hit enter, you get the red field and ?? everywhere as expected. But then if you switch to the browse tab you get the spinner and it never goes away, even when you switch back to the search tab. In other cases of illegal values when you switch to the browse tab you just get a blank gallery." What's worse you actually get a yellow "Message" in the top telling you there's an illegal value.
- This is still true in forms.py: "OK, I see what's going on here - since the min/max is also a class name. That does seem kind of like a bad class name, though. How about prepending something like "op-range-input-" to the "min"/"max" to make a better class name?"
- This question was not answered: "What happens with "intcomma" when we have non-int/float field values, like date/time strings?"
- The current JSON format doesn't actually support hint fields for the "max" field because it relies on there being a "slug2" column in the database.

juzen2003 committed:
Update the design to have "field_hints1" and "field_hints2" in param_info
table. The design will support hints for both min/max. We will just add
desired hints to "pi_field_hints1" and "pi_field_hints2" in slug/slug1
field in obs_xxx.json.
- In obs_ring_geometry.json (line 86-87) & obs_wavelength.json (line 83-84),
  add hints to "pi_field_hints1" and "pi_field_hints2".
- In paraminfo/models.py (line 28-29), add field_hints1 & field_hints2 fields.
- In do_param_info.py (line 106-107) and param_info.json (line 100-108),
  store values from "pi_field_hints1" and "pi_field_hints2" (obs_xxx.json)
  in "field_hints1" & "field_hints2" (database).
- In ui/views.py (line 300-301, 315-316), pass in "field_hints1" &
  "field_hints2" to widget.html.
- In search/forms.py (line 120-126), get the hints for slug2 from slug1
  param_info.

juzen2003 committed:
Remove unused field_hints1 & field_hints2 passed to widget.html.
- In ui/views.py (line 300, 312), remove field_hints1 & field_hints2
  passed to widget.html (they are not used in html).

juzen2003 committed:
Customize ring radius and wavelength dropdown responsiveness separately.
In wavelength, the names will never wrap. In ring radius dropdown, the
names will not wrap when there is still a lot of spaces left.
- In opus.css (line 1992-2042), customize ring radius and wavelength dropdown
  responsiveness.

juzen2003 committed:
Combine multiple if blocks into one single if block with && in search.js.
- In search.js (line 77-81), update the logic to combine multiple if blocks
  conditions into one single if block with &&.

juzen2003 committed:
Change dataimages api to call from getJSON to ajax, and add an error handler
to remove the spinner when bad search happened. This will remove the spinner
and leave gallery view blank.
- In browse.js (line 2096-2136), change getJSON to ajax and add an error
  handler to remove the spinner when dataimages api call has an error.

rfrenchseti reviewed (changes requested):
There are still problems with illegal search values:
- Type "thebe" into ring radius min and hit enter
- Switch to the browse tab
- Notice the hash now has "RINGGEOringradius1=thebe" in it. We should never allow illegal values in the hash.
- If you refresh the page now the illegal slug value will get passed to the backend and a yellow warning message will be shown.
- Go back, clear everything out and start over; then end "thebe" again
- Go to the browse results page and notice that you can drag the slider
- There are still cases where you go to the browse results tab with an illegal search value and it gives you a spinner

I think looking for the error from the AJAX call is the wrong way to handle this. We should already know that the input is invalid, right? Although there's probably an annoying race condition in there.

- The dropdown box is just a little too wide. If there's a vertical scrollbar in the search widget column, the dropdown will partially cover it.

juzen2003 committed:
Make sure ranges dropdown will not cover the widget scrollbar.
- In opus.css (line 1989-1990), add left values to tune the ranges dropdown
  location so that it will never overlap with the widget scrollbar.

juzen2003 committed:
Disable the browse tab nav link when there is an invalid search value in
range input. This is done by disabling browse tab nav link when user focuses
out and there is a change of value in range input. The button will be enabled
or keep disabled based on the result of input validation in
parseFinalNormalizedInputDataAndUpdateHash.
- In browse.js (line 2096-2129), switch ajax back to getJSON for dataimages
  api call.
- In search.js (line 90-100, 780, 701), in range input focusout event handler,
  disable browse tab nav link when user focuses out and there is a change
  of value in range input. The button will be enabled or keep disabled based
  on the result of input validation in parseFinalNormalizedInputDataAndUpdateHash.
- In opus.js (line 656-659), add click event handler to the disabled browse
  tab. This is to make sure browse tab nav link does nothing when it's
  been disabled.
- In hash.js (line 40). modify updateHash to make sure we don't update URL
  with invalid slug values.
- In base.html (line 14), add .op-browse-tab to browse nav link.
- In opus.css (line 2116-2120), add styling for .op-disabled-nav-link to
  make it not clickable.

juzen2003 committed:
Fix the issue that the disabled browse tab is not re-enabled when user closes
widgets with invalid value (rest of search values are valid).
- In widgets.js (line 216-225), when closing the widget, if all inputs are
  valid, we enable the browse tab. And if any input is invalid, make sure
  browse tab stays disabled.
- In opus.js (line 258-261), when getting the result count, properly
  disable/enable browse tab based on opus.allInputsValid.
